### PR TITLE
add make targets for building server images

### PR DESCRIPTION
**Pull request by BenTheElder**
**What this PR does / why we need it**: Adds `make release-images` and `make quick-release-images`, which allow building the docker-wrapped images without building a full release. Without these you can either use `make {quick}-release` and build test tarballs etc, or hack around in the build system yourself. Using this can be considerably faster if you just want to build the binaries and images, and not the release tarballs etc.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

/sig release

### CODE REVIEW

**Comment by BenTheElder** [comment]
A bit more context:
We're working on some tooling for local (dind) clusters for testing, and I've currently worked around this limitation so that it is possible to build a cluster using ~make / docker similar to bazel without building an entire release. Unfortunately I couldn't find a way to do this that didn't involve sourcing a bunch of our bash libs and calling some bash functions while making assumptions about the build, hence this.

Gross code for that is here: https://github.com/kubernetes/test-infra/blob/51c72d8129024d517c7544abf00ff8975b874d0c/kind/pkg/build/kube/dockerbuildbits.go

**Review by cblecker — COMMENTED** [review]
One comment. Otherwise looks fine to me. All net-new, so shouldn't impact existing make targets.

File: build/root/Makefile [hunk 1]
```diff
@@ -408,6 +426,27 @@ release-skip-tests quick-release:
...
+#
+# Example:
+#   make quick-release-images
+endef
+.PHONY: quick-release-images
+ifeq ($(PRINT_HELP),y)
+quick-release-images:
+	@echo "$$QUICK_RELEASE_IMAGES_HELP_INFO"
+else
+quick-release-images: KUBE_RELEASE_RUN_TESTS = n
```
[the remark this answers is not in the document]
Comment (BenTheElder) [reply build/root/Makefile]: fixed 🙄, thanks

File: build/lib/release.sh [hunk 2]
```diff
@@ -233,6 +233,31 @@ function kube::release::package_server_tarballs() {
...
+
+    local release_stage="${RELEASE_STAGE}/server/${platform_tag}/kubernetes"
+    rm -rf "${release_stage}"
+    mkdir -p "${release_stage}/server/bin"
+    mkdir -p "${release_stage}/addons"
+
+    # This fancy expression will expand to prepend a path
+    # (${LOCAL_OUTPUT_BINPATH}/${platform}/) to every item in the
+    # KUBE_SERVER_BINARIES array.
+    cp "${KUBE_SERVER_BINARIES[@]/#/${LOCAL_OUTPUT_BINPATH}/${platform}/}" \
```
Comment (BenTheElder) [inline build/lib/release.sh]: this CP is now wrong, I need to make it take the binaries we're actually wrapping which may not include hyperkube

File: build/lib/release.sh [hunk 3]
```diff
@@ -233,6 +233,31 @@ function kube::release::package_server_tarballs() {
   done
 }
 
+# Package up all of the server binaries in docker images
+function kube::release::build_server_images() {
```
[the remark this answers is not in the document]
Comment (ixdy) [reply build/lib/release.sh]: can we remove the part of `package_server_tarballs` that creates images and call this first?

File: build/lib/release.sh (hunk 2 above)
Comment (BenTheElder) [reply build/lib/release.sh]: fixed

File: build/lib/release.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (BenTheElder) [reply build/lib/release.sh]: it's in a loop though over the platforms that does some other things, do we want to loop over the platforms twice?

there's also some subtle difference in the target lists referenced by each, I preferred not doing anything possibly breaking to the existing functionality, only adding new

**Comment by BenTheElder** [comment]
/hold cancel
target lists for copying the binaries should be set now, as well as a special case for hyperkube

File: build/lib/release.sh [hunk 4]
```diff
@@ -233,6 +233,38 @@ function kube::release::package_server_tarballs() {
...
+
+    # This fancy expression will expand to prepend a path
+    # (${LOCAL_OUTPUT_BINPATH}/${platform}/) to every item in the
+    # KUBE_SERVER_IMAGE_BINARIES array.
+    cp "${KUBE_SERVER_IMAGE_BINARIES[@]/#/${LOCAL_OUTPUT_BINPATH}/${platform}/}" \
+      "${release_stage}/server/bin/"
+
+    # if we are building hyperkube, we also need to copy that binary
+    if [[ "${KUBE_BUILD_HYPERKUBE}" =~ [yY] ]]; then
+      local hyperkube="cmd/hyperkube"
```
Comment (BenTheElder) [inline build/lib/release.sh]: arg this needs to be` hyperkube` not `cmd/hyperkube`. I haven't actually been building hyperkube so this slipped

**Comment by BenTheElder** [comment]
Sorry for the noise, I should have done more testing previously with `KUBE_BUILD_HYPERKUBE=y`, I'd only been building with `KUBE_BUILD_HYPERKUBE=n`.

**Comment by spiffxp** [comment]
/milestone v1.12
SIG Testing chair here, this doesn't impact the existing release build path so I'm comfortable with it getting merged in this close to code freeze

File: build/lib/release.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (spiffxp) [reply build/lib/release.sh]: It's unclear to me what the cost is of iterating over platforms twice.  It looks like the subtle difference is `KUBE_SERVER_BINARIES` vs `KUBE_SERVER_IMAGE_BINARIES`?  It seems like you could de-dupe by having this function take its target list as a parameter and invoking accordingly.

**Review by spiffxp — COMMENTED** [review]
/lgtm
/hold
I am OK with merging this as-is if you can do a followup PR to de-dupe (that doesn't necessarily have to make it in before code freeze).  Remove the /hold when you've got a followup issue filed

File: build/lib/release.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (BenTheElder) [reply build/lib/release.sh]: I have a change for this locally, will hold pushing it until we see what @ixdy and @cblecker would prefer with doing this before or after code-freeze. I'm a little wary of touching the existing functionality so close to code freeze.

[the remark this answers is not in the document]
Comment (BenTheElder) [reply build/lib/release.sh]: discussed offline with @ixdy, adding another commit to do this refactor

File: build/lib/release.sh [hunk 5]
```diff
@@ -208,11 +210,38 @@ function kube::release::package_server_tarballs() {
 
     # This fancy expression will expand to prepend a path
```
Comment (ixdy) [inline build/lib/release.sh]: the `mkdir -p "${release_stage}/addons` line probably belongs in `package_server_tarballs`

File: build/lib/release.sh [hunk 6]
```diff
@@ -208,11 +210,38 @@ function kube::release::package_server_tarballs() {
...
-    # KUBE_SERVER_BINARIES array.
-    cp "${KUBE_SERVER_BINARIES[@]/#/${LOCAL_OUTPUT_BINPATH}/${platform}/}" \
+    # KUBE_SERVER_IMAGE_BINARIES array.
+    cp "${KUBE_SERVER_IMAGE_BINARIES[@]/#/${LOCAL_OUTPUT_BINPATH}/${platform}/}" \
       "${release_stage}/server/bin/"
 
+    # if we are building hyperkube, we also need to copy that binary
+    if [[ "${KUBE_BUILD_HYPERKUBE}" =~ [yY] ]]; then
+      local hyperkube="hyperkube"
+      cp "${hyperkube/#/${LOCAL_OUTPUT_BINPATH}/${platform}/}" \
```
Comment (ixdy) [inline build/lib/release.sh]: you can simplify this to just `cp "${LOCAL_OUTPUT_BINPATH}/${platform}/hyperkube" "${release_stage}/server/bin"` I think

Comment (BenTheElder) [reply build/lib/release.sh]: done

File: build/lib/release.sh (hunk 5 above)
Comment (BenTheElder) [reply build/lib/release.sh]: done

File: build/lib/release.sh [hunk 7]
```diff
@@ -193,17 +193,46 @@ function kube::release::package_node_tarballs() {
   done
 }
 
+# Package up all of the server binaries in docker images
+function kube::release::build_server_images() {
+  # Clean out any old images
+  rm -rf "${RELEASE_IMAGES}"
```
Comment (cblecker) [inline build/lib/release.sh]: @spiffxp @BenTheElder @ixdy Only concern is now we have hooked in this new code path with the existing release path, and this command in particular is new. If we have any hiccups with the release build, this command would be my first suspect. I don't foresee any issues, but this is the one difference I see.

Comment (BenTheElder) [reply build/lib/release.sh]: yeah, FWIW discussed this with ixdy and put the "refactor this new logic to be shared with the existing build" in a seperate commit, so we can revert that commit if necessary and still have the new functionality even. I was pretty wary to touch the existing release logic

Comment (BenTheElder) [reply build/lib/release.sh]: We also ran `pull-kubernetes-cross` on this PR and the output looks right. If anything does somehow break I'll immediately file a revert for the second commit, I was happy enough without it myself :-)

Comment (cblecker) [reply build/lib/release.sh]: SGTM. If it breaks, we blame Jeff. ;)

Comment (ixdy) [reply build/lib/release.sh]: bbl, taking a vacation until 1.12 is released.

Comment (spiffxp) [reply build/lib/release.sh]: Just remember, I was fine merging it as-is... 🚶

**Comment by spiffxp** [comment]
/hold cancel
as we have opted to go for the refactor now, no need to hold waiting for a followup issue

**Comment by BenTheElder** [comment]
:eyes: will keep an eye on things, also maybe starting cherry picks back to 1.11 / 1.10. thanks all!

**Pull request merged**